misc/multiaddr/src/lib.rs

```diff
@@ -303,6 +303,44 @@ impl TryFrom<Vec<u8>> for Multiaddr {
     }
 }
 
+/// Create a [`Multiaddr`] from the given IP address and suffix.
+pub fn ip_to_multiaddr(ip: IpAddr, suffix: &[Protocol]) -> Multiaddr {
+    let proto = match ip {
+        IpAddr::V4(ip) => Protocol::Ip4(ip),
+        IpAddr::V6(ip) => Protocol::Ip6(ip),
+    };
+    let it = std::iter::once(proto).chain(suffix.into_iter().cloned());
+    Multiaddr::from_iter(it)
+}
+
+/// Collect all local host addresses and use the provided port number as listen port.
+#[cfg(not(any(target_os = "emscripten", target_os = "unknown")))]
+pub fn host_addresses(suffix: &[Protocol]) -> io::Result<Vec<(IpAddr, ipnet::IpNet, Multiaddr)>> {
```

Comment:
I understand that you're trying to reduce code duplication, but right now the `multiaddr` crate is a no-std-friendly crate that doesn't rely on any feature of the operating system. I'm not sure that it's worth breaking this assumption.
On the other hand, I don't really know where to put this function.